daemons/controld/controld_transition.c

```diff
@@ -186,6 +186,7 @@ do_te_invoke(long long action,
                  controld_globals.transition_graph->id, ref, graph_input);
 
         te_reset_job_counts();
+        controld_set_fsa_input_flags(R_IN_TRANSITION);
```

Comment:
Timers:
* The finalization timeout defaults to 30min, so we don't necessarily need to do anything there.
```
    {
        "join-finalization-timeout", "crmd-finalization-timeout", "time", NULL,
        "30min", pcmk__valid_interval_spec,
        "*** Advanced Use Only ***",
        N_("If you need to adjust this value, it probably indicates "
            "the presence of a bug.")
    }
```
* When we entered the `S_FINALIZE_JOIN` state from `S_INTEGRATED` state, we stopped four other timers: (current) election, integration, election (trigger), and recheck:
```
static void
do_state_transition(enum crmd_fsa_state cur_state,
                    enum crmd_fsa_state next_state, fsa_data_t *msg_data)
{
...
    if (next_state != S_ELECTION && cur_state != S_RELEASE_DC) {
        controld_stop_current_election_timeout();
    }
#if 0
    if ((controld_globals.fsa_input_register & R_SHUTDOWN)) {
        controld_set_fsa_action_flags(A_DC_TIMER_STOP);
    }
#endif
    if (next_state == S_INTEGRATION) {
        controld_set_fsa_action_flags(A_INTEGRATE_TIMER_START);
    } else {
        controld_set_fsa_action_flags(A_INTEGRATE_TIMER_STOP);
    }

    if (next_state == S_FINALIZE_JOIN) {
        controld_set_fsa_action_flags(A_FINALIZE_TIMER_START);
    } else {
        controld_set_fsa_action_flags(A_FINALIZE_TIMER_STOP);
    }

    if (next_state != S_PENDING) {
        controld_set_fsa_action_flags(A_DC_TIMER_STOP);
    }
    if (next_state != S_IDLE) {
        controld_stop_recheck_timer();
    }
```
* There are three remaining timers:
    * `wait_timer` gets started only by `do_cib_control()` on `A_CIB_(STOP|START)|O_CIB_RESTART`, as part of a connection retry process. It only gets stopped by `crm_timer_popped()`.
    * `shutdown_escalation_timer` gets started only by `crm_shutdown()` and stopped by `crm_timer_popped()`.
    * `transition_timer` gets (re)started only by `notify_crmd()` and `abort_transition_graph()`. In any case, since we're delaying the join sequence to allow a transition to complete, the transition timer is probably taken care of.

Reply:
I think worst case scenario, if the finalization timer pops because the transition takes more than 30 minutes, we start a new election (see `finalization_timer->fsa_input`).

Reply:
Excellent, I think we're covered